``FreeGroup`` implementation in SymPy

gxyd opened the pull request:
Implementation of `FreeGroups` in SymPy, the way in which the output is produced resonates very much with the way in `GAP`.
1. I have not added the tests. But the doctests would serve the purpose, since they pass the tests.
2. Some of the important things like `subgroup` still need to be added and more infact..
3. I closed the previous PR #10263 since i think this PR's approach looks good to me.(i did worse in previous PR to even think of using `symbols` at some places i think).

ping @jksuom

Commit:
``FreeGroup`` implementation in SymPy

jksuom commented on sympy/combinatorics/free_group.py:
```diff
@@ -0,0 +1,918 @@
+
+            # (2) form of the API used here
+            elif len(args) == 2:
+                obj._str = args[1]
```
Should the type of ``args[1]`` be checked?

gxyd commented:
Also one thing that currently does not work is the way simplification occurs in ``GAP``.
Currently
```
>>> f = FreeGroup( 4 )
>>> (f[0]*f[1])**2
f[0]*f[1]*f[0]*f[1]         # better represented as (f[0]*f[1])**2  i.e no expansion
```
I will next try to see this point.

jksuom commented on sympy/combinatorics/free_group.py:
```diff
@@ -0,0 +1,918 @@
+        if self is other:
+            return True
+        else:
+            return False
```
Perhaps ``return self is other`` would suffice, and it is probable that Python would provide this as a default. (From the [library manual](https://docs.python.org/2/library/stdtypes.html#comparisons): "Non-identical instances of a class normally compare as non-equal unless the class defines the ``__eq__()`` method or the ``__cmp__()`` method.")

jksuom commented on sympy/combinatorics/free_group.py:
```diff
@@ -0,0 +1,918 @@
+        Examples
+        ========
+        """
+        if F.is_group and set(F.generators).issubset(set(self.generators)):
```
I think this should be something like ``all([self.contains(gen) for gen in F.generators])``.

jksuom commented on sympy/combinatorics/free_group.py:
```diff
@@ -0,0 +1,918 @@
+    def array_form(self):
+        """
+        SymPy provides two different internal kinds of representation
+        of assosciative words. The first one is called the `array_form`
```
typo: assosciative

Commit:
code simplified

gxyd commented on sympy/combinatorics/free_group.py:
```diff
@@ -0,0 +1,916 @@
+        True
+
+        """
+        return self is other
```
> it is probable that Python would provide this as a default

But for the above doctest i have added. It is such that even if the two instances `f` and `g` have the same arguments (i.e "4" and "swapnil") but still `f == g` has to be returned as being not equal, different from the usual way Python defines `__eq__`(which returns `True` when the arguments are same for two instances of same class).

jksuom replied:
I see... This behaviour is probably inherited from ``Basic``. I wonder if it is necessary to make ``Group`` a subclass of ``Basic``. (``PolyRing`` is not...) @asmeurer ?

asmeurer replied:
Why do you want this behavior to work this way? The whole idea behind `Basic.__eq__` is that == works as structural (not mathematical) equality, which tends to be the most useful thing when dealing with symbolic objects.

gxyd replied:
Then, did we made a mistake [here](https://github.com/sympy/sympy/pull/10230/files#diff-09b92940e0335c06b640b110d8549cf6R191) in PR #10230 for the `PermutationGroup`?

Commit:
math fixes in code

asmeurer commented:
I'm not a fan of string based APIs. It's better to create a GroupSymbol object, than to use strings. Otherwise you will always have annoying indirection, since you can't actually multiply strings together.

asmeurer commented on sympy/combinatorics/free_group.py:
```diff
@@ -0,0 +1,916 @@
+    generators as the length of the list `names`, the i-th generator being
+    printed as `names[i]`.
+
+    ``FreeGroup( "string0", "string1", "string2", ... )`` ..............  (3)
```
Is this literally with an ellipsis literal (in Python 3)?

gxyd replied:
No, this is not. I will remove this.

gxyd commented:
> I'm not a fan of string based APIs.

I too thought the same, hence i left the work for 3rd API form that i have mentioned.

Commit:
Merge branch 'master' into implementation_FreeGroup

jksuom commented on sympy/combinatorics/free_group.py:
```diff
@@ -0,0 +1,916 @@
+        return obj
+
+    @property
+    def group(self):
```
Is there some reason why this could not be an attribute? Maybe even a class attribute.

gxyd replied:
This attribute of `group` will vary according to the Free Group element, each `FreeGroupElm` will have a different Free Group, which will be provided to `__new__(cls, free_group, ... )` just above this code.

gxyd replied:
This can be an instance attribute, but not a class attribute.

jksuom commented on sympy/combinatorics/free_group.py:
```diff
@@ -0,0 +1,916 @@
+def zero_simp(array_form):
+    for i in range(len(array_form) - 1, -1, -1):
+        if array_form[i][1] == 0:
+            array_form[i: i + 1] = []
```
There is also `del array_form[i]` that I find more readable.

jksuom commented on sympy/combinatorics/free_group.py:
```diff
@@ -0,0 +1,916 @@
+                l.append(tuple(app))
+        zero_simp(l)
+        mult_simp(l)
+        zero_simp(l)
```
After the second `zero_simp(l)` another `mult_simp(l)` may be needed.  Perhaps those functions should be combined.

Commit:
remove un-necessary import

Commit:
trying to remove the __new__ method from FreeGroupElm

Commit:
make FreeGroupElm a subclass of list instead of basic

Commit:
changes in methods using dtype

Commit:
Add test file for free_group.py

gxyd commented:
> I'm not a fan of string based APIs. It's better to create a GroupSymbol object, than to use strings. Otherwise you will always have annoying indirection, since you can't actually multiply strings together.

Though it is not fully developed at all. I am also in favour of removing that. Now, should i remove that? (this comment was made a while back, so making sure no un-necessary changes). This will change API of `FreeGroup` from `*args, **kwargs` to `rank, str_expr`.

Commit:
checking the input type with `isint`

Commit:
zero_mul_simp method improved

jksuom commented on sympy/combinatorics/free_group.py:
```diff
@@ -0,0 +1,872 @@
+from sympy.core.compatibility import as_int
+from sympy.core.sympify import CantSympify
+from mpmath import isint
+from collections.abc import Sequence
```
`collections.abc` does not seem to exist in 2.7. There is plain `abc`, but it does not contain `Sequence`. Maybe this is not necessary.

Commit:
change free group API to similar to polys

Commit:
Add `tuple` as baseclass for FreeGroupElm not list

Commit:
changes in zero_mul_simp function

jksuom commented on sympy/combinatorics/free_group.py:
```diff
@@ -116,11 +110,11 @@ def generators(self):
         [swapnil0, swapnil1, swapnil2]
 
         """
+        group = self
```
I believe this could be omitted if you write ``def _generators(group):`.

Commit:
printing and different free group introduction

Commit:
changes in methods

Commit:
documentation and (index, exp) -> (symbol, exp) changes

jksuom commented on sympy/combinatorics/free_group.py:
```diff
@@ -118,15 +118,15 @@ def _generators(group):
 
         """
         gens = []
         for i in range(group.rank):
```
Maybe `for sym in group.symbols:`.

jksuom commented:
Looks good. I commented the changes, but it seems to have become 'outdated' (at least in my browser).

Commit:
remove the unused import of collections.abc

gxyd commented:
I have the removed the unused import of `collections.abc` :smile:

Commit:
changes in methods and __truediv__

Commit:
comparison with `is` instead of `==` for Symbol

Commit:
base class changed from Basic -> DefaultPrinting

Commit:
hash method for FreeGroup and FreeGroupElm and _parse_symbols

Commit:
tests files initialised with good tests

Commit:
add tests and comparison of FreeGroupElm fixed

Commit:
add tests for "words", "syllables"

gxyd commented:
I have added the tests, added a few more `methods`. Also the `_parse_symbols` is now directly used only by the `FreeGroup` class, instead of being used individually by the `vfree_group` or `free_group` etc.
There are a few changes that result due to the use of `(symbol, exp)` instead of `(index, exp)` , this include `index + some_number` gives the next `generator` while so is not the case with `(symbol, exp)`.
Only a few more tests are needed.

jksuom commented on sympy/combinatorics/free_group.py:
```diff
@@ -287,6 +293,14 @@ class FreeGroupElm(CantSympify, tuple):
+
+    def __hash__(self):
+        _hash = self._hash
+        if _hash is None:
```
Can this really happen?

gxyd replied:
When the an instance of `FreeGroupElm` is created, initially it has `_hash = None` (see the assignment just above this), so `_hash is None` will be True. I don't see anything bad with this check.

jksuom replied:
Neither do I. Actually, I did not notice that this was for `FreeGroupElm` and not for `FreeGroup`.

Commit:
Fix docstring, printing issue of FreeGroupElm

gxyd commented:
I have made some docstring, printing fixes. The printing fix is a little hacky, though. Apart from that, I think docstrings are complete now.

Commit:
make lw=len(array_form), remove while loop

Commit:
use loop instead of recursion for zero_mul_simp

Commit:
change the zero_simp_mul method

Commit:
Raise TypeError, Add tests for invalid FreeGroupElm operations

Commit:
remove useless import

gxyd commented:
I think this is now ready for a final review.

Commit:
change condition from len(l) >1 to index>=0 in zero_mul_simp

jksuom commented:
I think this is ready to be merged.

gxyd commented:
ping @jksuom @asmeurer . Let me know before anyone merges this. I will squash the commits. Next will start with "FreeAbelianGroup", "Coset Enumeration".

asmeurer commented on sympy/combinatorics/free_group.py:
```diff
@@ -0,0 +1,938 @@
+    return (_free_group,) + tuple(_free_group.generators)
+
+@public
+def xfree_group(symbols):
```
I know this is already merged, but what's the point of this separate function?

gxyd replied:
specifically the `xfree_group` could be used for collecting the generators of free group in a single variable, directly. Similarly the next function `vfree_group` could be use to pollute the environment with the generator variables automatically. I don't say we can't live without it. But its good to have it.

asmeurer commented on sympy/combinatorics/free_group.py:
```diff
@@ -0,0 +1,938 @@
+#                          FREE GROUP                                        #
+##############################################################################
+
+_free_group_cache = {}
```
Why do you need this, and why don't you use the SymPy cache. Unbounded caches can be problematic.

asmeurer commented on sympy/combinatorics/free_group.py:
```diff
@@ -0,0 +1,938 @@
+    def __eq__(self, other):
+        """No ``FreeGroup`` is equal to any "other" ``FreeGroup``.
+        """
+        return self is other
```
Why this? It seems the usual SymPy definition of `__eq__` would be more useful.

gxyd replied:
This could be changed. Its not something under a priority, so I left it as it is.

asmeurer commented on sympy/combinatorics/free_group.py:
```diff
@@ -0,0 +1,938 @@
+    __rtruediv__ = __rdiv__
+
+    def __add__(self, other):
+        raise TypeError("unsupported operand type(s) for + or add: '%s' and '%s'"
```
The correct behavior here (and possibly in other cases where you're raising `TypeError`) is to `return NotImplemented`. That way, other classes can define `__add__` and `__radd__` to add FreeGroupElm objects.

gxyd replied:
Addressed in "coset enumeration" PR #11140 .

asmeurer commented on sympy/combinatorics/free_group.py:
```diff
@@ -0,0 +1,938 @@
+############################################################################
+
+
+class FreeGroupElm(CantSympify, DefaultPrinting, tuple):
```
I would spell out `FreeGroupElement`.

gxyd replied:
That would be fine.

gxyd replied:
Addressed in "coset enumeration" PR #11140 .

asmeurer commented on sympy/printing/pretty/pretty.py:
```diff
@@ -1715,6 +1715,9 @@ def _print_PolyRing(self, ring):
         return prettyForm(sstr(field))
 
+    def _print_FreeGroupElm(self, elm):
+        return prettyForm(str(elm))
```
Actual pretty printing would be nice. But for that, you may need to be using Expr or Basic subclasses.
